File: native-ios/MGLFleetSDK/Sources/MGLFleetSDK/DriverAppApiClient.swift
import Foundation

/// Live driver-app HTTP; mirrors `driver-api.ts` / Android `DriverAppApiClient`. Use when `!FleetSdkOptions.useMock`.
final class DriverAppApiClient {
    private let options: FleetSdkOptions
    private let session: URLSession

    private static let userAgent = "mgl-fleet-ios-sdk/1.0"

    init(options: FleetSdkOptions, session: URLSession = .shared) {
        self.options = options
        self.session = session
    }

    private var baseTrimmed: String {
        options.apiBaseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    }

    private func unwrapBody(from text: String) throws -> Any? {
        let parsed = try DriverFleetJSON.parseFlexible(text)
        return try DriverFleetJSON.unwrapDriverBody(parsed)
    }

    private func getRaw(path: String, bearer: String?) async throws -> String {
        guard let url = URL(string: baseTrimmed + path) else { throw DriverApiError.message("Bad URL") }
        var req = URLRequest(url: url)
        req.httpMethod = "GET"
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        req.setValue(Self.userAgent, forHTTPHeaderField: "User-Agent")
        if let b = bearer?.trimmingCharacters(in: .whitespaces), !b.isEmpty {
            req.setValue("Bearer \(b)", forHTTPHeaderField: "Authorization")
        }
        let (data, resp) = try await session.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw DriverApiError.message("No HTTP response") }
        let text = String(data: data, encoding: .utf8) ?? ""
        if !(200 ..< 300).contains(http.statusCode) {
            let msg =
                (
                    try? DriverFleetJSON.extractFleetApiErrorMessage(try DriverFleetJSON.parseFlexible(text)))
                ?? nil
            throw DriverApiError.message(msg ?? "HTTP \(http.statusCode)")
        }
        return text
    }

    private func postJsonRaw(path: String, body: [String: Any], bearer: String?) async throws -> String {
        guard let url = URL(string: baseTrimmed + path) else { throw DriverApiError.message("Bad URL") }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        req.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        req.setValue(Self.userAgent, forHTTPHeaderField: "User-Agent")
        if let b = bearer?.trimmingCharacters(in: .whitespaces), !b.isEmpty {
            req.setValue("Bearer \(b)", forHTTPHeaderField: "Authorization")
        }
        req.httpBody =
            try JSONSerialization.data(withJSONObject: body)
        let (data, resp) = try await session.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw DriverApiError.message("No HTTP response") }
        let text = String(data: data, encoding: .utf8) ?? ""
        if !(200 ..< 300).contains(http.statusCode) {
            let msg =
                (
                    try? DriverFleetJSON.extractFleetApiErrorMessage(try DriverFleetJSON.parseFlexible(text)))
                ?? nil
            throw DriverApiError.message(msg ?? "HTTP \(http.statusCode)")
        }
        return text
    }

    func driverCheckMobile(_ mobile: String) async -> Result<CheckMobileStatusKind, Error> {
        await runCatching {
            let txt = try await postJsonRaw(path: "/api/v0/driver-app/auth/check-mobile", body: ["mobile": mobile], bearer: nil)
            let unwrapped = try unwrapBody(from: txt)
            return try DriverFleetJSON.parseCheckMobile(unwrapped)
        }
    }

    func driverSendLoginOtp(_ mobile: String) async -> Result<Void, Error> {
        await runCatching {
            let enc = mobile.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? mobile
            _ = try await getRaw(path: "/api/v0/otp/login?username=\(enc)", bearer: nil)
        }
    }

    func driverInviteMobileSendOtp(_ mobile: String) async -> Result<String, Error> {
        await runCatching {
            let txt = try await postJsonRaw(path: "/api/v0/driver-app/auth/mobile/send-otp", body: ["mobile": mobile], bearer: nil)
            let parsed = try DriverFleetJSON.parseFlexible(txt)
            return try DriverFleetJSON.unwrapInviteSendOtpInner(parsed)
        }
    }

    func driverInviteMobileVerifyOtp(mobile: String, otpRefNumber: String, otp: String) async -> Result<String, Error> {
        await runCatching {
            let txt =
                try await postJsonRaw(
                    path: "/api/v0/driver-app/auth/mobile/verify-otp",
                    body: [
                        "mobile": mobile,
                        "otpRefNumber": otpRefNumber,
                        "otp": otp,
                    ],
                    bearer: nil,
                )
            let parsed = try DriverFleetJSON.parseFlexible(txt)
            let unwrappedAny = try DriverFleetJSON.unwrapDriverBody(parsed)
            let jo = try DriverFleetJSON.requireJSONObject(unwrappedAny)
            guard let tok = jo["mobileVerificationToken"] as? String, !tok.trimmingCharacters(in: .whitespaces).isEmpty else {
                throw DriverApiError.message("Unexpected verify-otp response")
            }
            return tok
        }
    }

    func driverInviteValidate(mobile: String, inviteCode: String, mobileVerificationToken: String) async -> Result<InviteValidateParsed, Error> {
        await runCatching {
            let txt =
                try await postJsonRaw(
                    path: "/api/v0/driver-app/auth/invite/validate",
                    body: [
                        "mobile": mobile,
                        "inviteCode": inviteCode,
                        "mobileVerificationToken": mobileVerificationToken,
                    ],
                    bearer: nil,
                )
            let unwrapped = try unwrapBody(from: txt)
            let jo = try DriverFleetJSON.requireJSONObject(unwrappedAny(unwrapped))
            return try DriverFleetJSON.parseInviteValidate(jo)
        }
    }

    /// After JSON decode, unwrap string-wrapped payloads for OAuth bodies.
    private func unwrappedAny(_ any: Any?) -> Any? {
        if let s = any as? String, let d = s.data(using: .utf8),
           let o = try? JSONSerialization.jsonObject(with: d) {
            return o
        }
        return any
    }

    func driverInviteSetPin(sessionToken: String, pin: String) async -> Result<String, Error> {
        await runCatching {
            let txt =
                try await postJsonRaw(
                    path: "/api/v0/driver-app/auth/invite/set-pin",
                    body: ["sessionToken": sessionToken, "pin": pin],
                    bearer: nil,
                )
            let parsed = try DriverFleetJSON.parseFlexible(txt)
            guard let uw = try DriverFleetJSON.unwrapDriverBody(parsed) else { throw DriverApiError.message("invite set-pin: empty body") }
            if let tok = DriverFleetJSON.oauthAccessToken(unwrappedAny(uw)) { return tok }
            if let inner = uw as? [String: Any], let nested = DriverFleetJSON.oauthAccessToken(inner) { return nested }
            throw DriverApiError.message("Missing access token")
        }
    }

    func driverOauthOtpGrant(mobile: String, otp: String) async -> Result<String, Error> {
        await runCatching {
            guard let url = URL(string: baseTrimmed + "/oauth/token") else { throw DriverApiError.message("Bad oauth URL") }
            var req = URLRequest(url: url)
            req.httpMethod = "POST"
            req.setValue("application/json", forHTTPHeaderField: "Accept")
            req.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
            req.setValue(Self.userAgent, forHTTPHeaderField: "User-Agent")
            let pairs: [(String, String)] = [
                ("grant_type", "otp"),
                ("username", mobile.trimmingCharacters(in: .whitespaces)),
                ("otp", otp.trimmingCharacters(in: .whitespaces)),
                ("client_id", "mgl-driver-app-client"),
                ("client_secret", "driver-app-secret"),
            ]
            let body =
                pairs
                    .map { kv in
                        let k = kv.0.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? kv.0
                        let v = kv.1.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? kv.1
                        return "\(k)=\(v)"
                    }
                    .joined(separator: "&")
            req.httpBody = body.data(using: .utf8)

            let (data, resp) = try await session.data(for: req)
            guard let http = resp as? HTTPURLResponse else { throw DriverApiError.message("No HTTP response") }
            let text = String(data: data, encoding: .utf8) ?? ""
            if !(200 ..< 300).contains(http.statusCode) {
                let parsed = try? DriverFleetJSON.parseFlexible(text)
                throw DriverApiError.message(DriverFleetJSON.peelOAuthError(parsed))
            }
            let any = try DriverFleetJSON.parseFlexible(text)
            guard let jo = any as? [String: Any],
                  let t = jo["access_token"] as? String ?? jo["accessToken"] as? String,
                  !t.isEmpty else { throw DriverApiError.message("OAuth: no token") }
            return t
        }
    }

    func driverFoList(_ bearerPartial: String) async -> Result<[FoListEntryParsed], Error> {
        await runCatching {
            let txt = try await getRaw(path: "/api/v0/driver-app/auth/fo-list", bearer: bearerPartial)
            let uw = try unwrapBody(from: txt)
            return DriverFleetJSON.parseFoList(uw)
        }
    }

    func driverFoSelect(bearerPartial: String, foCompanyId: Int64, pin: String) async -> Result<String, Error> {
        await runCatching {
            let txt =
                try await postJsonRaw(
                    path: "/api/v0/driver-app/auth/fo-select",
                    body: ["foCompanyId": NSNumber(value: foCompanyId), "pin": pin],
                    bearer: bearerPartial,
                )
            let parsed = try DriverFleetJSON.parseFlexible(txt)
            guard let uw = try DriverFleetJSON.unwrapDriverBody(parsed) else { throw DriverApiError.message("fo-select empty") }
            if let tok = DriverFleetJSON.oauthAccessToken(unwrappedAny(uw)) { return tok }
            if let inner = uw as? [String: Any], let nested = DriverFleetJSON.oauthAccessToken(inner) { return nested }
            throw DriverApiError.message("Missing fleet token")
        }
    }

    func driverGetHome(_ token: String) async -> Result<DriverHomeParsed, Error> {
        await runCatching {
            let txt = try await getRaw(path: "/api/v0/driver-app/home", bearer: token)
            let uw = try unwrapBody(from: txt)
            let jo = try DriverFleetJSON.requireJSONObject(uw)
            return DriverFleetJSON.parseHome(jo)
        }
    }

    func driverGetProfile(_ token: String) async -> Result<DriverProfileParsed, Error> {
        await runCatching {
            let txt = try await getRaw(path: "/api/v0/driver-app/profile", bearer: token)
            let uw = try unwrapBody(from: txt)
            let jo = try DriverFleetJSON.requireJSONObject(uw)
            return DriverFleetJSON.parseProfile(jo)
        }
    }

    func driverGetAssignments(_ token: String) async -> Result<[DriverAssignmentParsed], Error> {
        await runCatching {
            let txt = try await getRaw(path: "/api/v0/driver-app/assignments", bearer: token)
            let uw = try unwrapBody(from: txt)
            return DriverFleetJSON.parseAssignments(uw)
        }
    }

    func driverAcceptPairing(_ token: String, pairingCode: String) async -> Result<(vrn: String, status: String), Error> {
        await runCatching {
            let txt =
                try await postJsonRaw(
                    path: "/api/v0/driver-app/vehicle/accept-pairing",
                    body: ["pairingCode": pairingCode],
                    bearer: token,
                )
            let uw = try unwrapBody(from: txt)
            let jo = try DriverFleetJSON.requireJSONObject(uw)
            let vrn = ((jo["vehicleRegNo"] as? String) ?? "").trimmingCharacters(in: .whitespaces)
            let status = ((jo["status"] as? String) ?? "").trimmingCharacters(in: .whitespaces)
            return (vrn, status)
        }
    }

    func driverQrPay(
        _ token: String,
        txnId: String,
        vehicleRegNoNorm: String,
        pin: String,
        mid: String,
        terminalId: String,
        amountPaise: Int64,
        expiryEpoch: Int64,
        sign: String,
    ) async -> Result<QrPayResultParsed, Error> {
        await runCatching {
            let txt =
                try await postJsonRaw(
                    path: "/api/v0/driver-app/qr/pay",
                    body: [
                        "txnId": txnId,
                        "vehicleRegNo": vehicleRegNoNorm,
                        "pin": pin,
                        "mid": mid,
                        "terminalId": terminalId,
                        "amountPaise": NSNumber(value: amountPaise),
                        "expiryEpoch": NSNumber(value: expiryEpoch),
                        "sign": sign,
                    ],
                    bearer: token,
                )
            let uw = try unwrapBody(from: txt)
            let jo = try DriverFleetJSON.requireJSONObject(uw)
            return DriverFleetJSON.parseQrPayResult(jo)
        }
    }

    func driverGetTransactions(_ token: String, vehicleId: String, page: Int = 0) async -> Result<TxnsPageParsedIOS, Error> {
        await runCatching {
            let vid = vehicleId.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? vehicleId
            let q = page > 0 ? "?page=\(page)" : ""
            let txt = try await getRaw(path: "/api/v0/driver-app/vehicles/\(vid)/transactions\(q)", bearer: token)
            let uw = try unwrapBody(from: txt)
            return DriverFleetJSON.normalizeDriverTransactionsPayload(uw)
        }
    }

    private func runCatching<T>(_ body: () async throws -> T) async -> Result<T, Error> {
        do {
            return .success(try await body())
        } catch let e as DriverApiError {
            return .failure(e)
        } catch {
            return .failure(error)
        }
    }
}

File: native-android/fleet-sdk/src/main/java/com/mgl/fleet/sdk/internal/DriverAppApiClient.kt
package com.mgl.fleet.sdk.internal

import com.mgl.fleet.sdk.FleetSdkOptions
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.FormBody
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

/** Live driver-app HTTP; mirrors `driver-api.ts`. Use when `!FleetSdkOptions.useMock`. */
internal class DriverAppApiClient(
    private val opts: FleetSdkOptions,
    private val client: OkHttpClient =
        OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .build(),
) {
    private val base: String get() = opts.apiBaseUrl.trimEnd('/')

    suspend fun driverCheckMobile(mobile: String): Result<CheckMobileStatus> =
        runCatchingSuspend {
            parseCheckMobile(requireJsonBody(postJsonRaw("/api/v0/driver-app/auth/check-mobile", JSONObject().put("mobile", mobile), bearer = null)))
        }

    suspend fun driverSendLoginOtp(mobile: String): Result<Unit> =
        runCatchingSuspend {
            val q = URLEncoder.encode(mobile, Charsets.UTF_8.name())
            getRaw("/api/v0/otp/login?username=$q", bearer = null)
            Unit
        }

    suspend fun driverInviteMobileSendOtp(mobile: String): Result<String> =
        runCatchingSuspend {
            unwrapInviteSendOtpInner(
                requireJsonBody(postJsonRaw("/api/v0/driver-app/auth/mobile/send-otp", JSONObject().put("mobile", mobile), bearer = null)),
            )
        }

    suspend fun driverInviteMobileVerifyOtp(
        mobile: String,
        otpRefNumber: String,
        otp: String,
    ): Result<String> =
        runCatchingSuspend {
            val body =
                JSONObject(
                    mapOf(
                        "mobile" to mobile,
                        "otpRefNumber" to otpRefNumber,
                        "otp" to otp,
                    ),
                )
            val o = requireJsonObject(unwrapDriverBodyJson(postJsonRaw("/api/v0/driver-app/auth/mobile/verify-otp", body, bearer = null)))
            val tok = o.optString("mobileVerificationToken", "").trim()
            if (tok.isEmpty()) throw DriverApiException("Unexpected verify-otp response")
            tok
        }

    suspend fun driverInviteValidate(
        mobile: String,
        inviteCode: String,
        mobileVerificationToken: String,
    ): Result<InviteValidateResult> =
        runCatchingSuspend {
            val body =
                JSONObject(
                    mapOf(
                        "mobile" to mobile,
                        "inviteCode" to inviteCode,
                        "mobileVerificationToken" to mobileVerificationToken,
                    ),
                )
            val o = requireJsonObject(unwrapDriverBodyJson(postJsonRaw("/api/v0/driver-app/auth/invite/validate", body, bearer = null)))
            InviteValidateResult(
                sessionToken = o.getString("sessionToken"),
                driverName = o.optString("driverName", ""),
                foName = o.optString("foName", ""),
                foCompanyId = o.optLong("foCompanyId"),
            )
        }

    suspend fun driverInviteSetPin(sessionToken: String, pin: String): Result<String> =
        runCatchingSuspend {
            val body = JSONObject(mapOf("sessionToken" to sessionToken, "pin" to pin))
            val unwrapped = unwrapDriverBodyJson(postJsonRaw("/api/v0/driver-app/auth/invite/set-pin", body, bearer = null))
            oauthAccessTokenFromAny(parseJsonFlexibleObj(unwrapped))
                ?: oauthAccessTokenFromAny(unwrapped)
                ?: throw DriverApiException("Missing access token")
        }

    suspend fun driverOauthOtpGrant(mobile: String, otp: String): Result<String> =
        withContext(Dispatchers.IO) {
            try {
                val form =
                    FormBody.Builder(Charsets.UTF_8)
                        .add("grant_type", "otp")
                        .add("username", mobile.trim())
                        .add("otp", otp.trim())
                        .add("client_id", "mgl-driver-app-client")
                        .add("client_secret", "driver-app-secret")
                        .build()
                val req =
                    Request.Builder()
                        .url("$base/oauth/token")
                        .post(form)
                        .header("Accept", "application/json")
                        .header("User-Agent", USER_AGENT)
                        .build()
                client.newCall(req).execute().use { resp ->
                    val text = resp.body?.string().orEmpty()
                    if (!resp.isSuccessful) {
                        val parsed = try { parseJsonFlexible(text) } catch (_: Exception) { null }
                        return@withContext Result.failure(DriverApiException(peelOAuthError(parsed)))
                    }
                    val jo = JSONObject(text)
                    val token =
                        jo.optString("access_token", "").takeIf { it.isNotEmpty() }
                            ?: jo.optString("accessToken", "").takeIf { it.isNotEmpty() }
                    if (token == null) Result.failure(DriverApiException("OAuth: no token"))
                    else Result.success(token)
                }
            } catch (e: DriverApiException) {
                Result.failure(e)
            } catch (e: Exception) {
                Result.failure(e)
            }
        }

    suspend fun driverFoList(bearerPartial: String): Result<List<FoListEntry>> =
        runCatchingSuspend {
            val data = unwrapDriverBodyJson(getRaw("/api/v0/driver-app/auth/fo-list", bearerPartial))
            foListFromAny(data)
        }

    suspend fun driverFoSelect(
        bearerPartial: String,
        foCompanyId: Long,
        pin: String,
    ): Result<String> =
        runCatchingSuspend {
            val body = JSONObject(mapOf("foCompanyId" to foCompanyId, "pin" to pin))
            val unwrapped = unwrapDriverBodyJson(postJsonRaw("/api/v0/driver-app/auth/fo-select", body, bearer = bearerPartial))
            oauthAccessTokenFromAny(parseJsonFlexibleObj(unwrapped))
                ?: oauthAccessTokenFromAny(unwrapped)
                ?: throw DriverApiException("Missing fleet token")
        }

    suspend fun driverGetHome(token: String): Result<DriverHomeJson> =
        runCatchingSuspend { parseHome(requireJsonObject(unwrapDriverBodyJson(getRaw("/api/v0/driver-app/home", token)))) }

    suspend fun driverGetProfile(token: String): Result<DriverProfileJson> =
        runCatchingSuspend { parseProfile(requireJsonObject(unwrapDriverBodyJson(getRaw("/api/v0/driver-app/profile", token)))) }

    suspend fun driverGetAssignments(token: String): Result<List<DriverAssignmentJson>> =
        runCatchingSuspend { parseAssignments(unwrapDriverBodyJson(getRaw("/api/v0/driver-app/assignments", token))) }

    suspend fun driverAcceptPairing(token: String, pairingCode: String): Result<Pair<String, String>> =
        runCatchingSuspend {
            val o =
                requireJsonObject(
                    unwrapDriverBodyJson(
                        postJsonRaw(
                            "/api/v0/driver-app/vehicle/accept-pairing",
                            JSONObject().put("pairingCode", pairingCode),
                            bearer = token,
                        ),
                    ),
                )
            o.optString("vehicleRegNo", "") to o.optString("status", "")
        }

    suspend fun driverQrPay(
        token: String,
        txnId: String,
        vehicleRegNoNorm: String,
        pin: String,
        mid: String,
        terminalId: String,
        amountPaise: Long,
        expiryEpoch: Long,
        sign: String,
    ): Result<QrPayResultJson> =
        runCatchingSuspend {
            val body =
                JSONObject(
                    mapOf(
                        "txnId" to txnId,
                        "vehicleRegNo" to vehicleRegNoNorm,
                        "pin" to pin,
                        "mid" to mid,
                        "terminalId" to terminalId,
                        "amountPaise" to amountPaise,
                        "expiryEpoch" to expiryEpoch,
                        "sign" to sign,
                    ),
                )
            val o = requireJsonObject(unwrapDriverBodyJson(postJsonRaw("/api/v0/driver-app/qr/pay", body, bearer = token)))
            val stRaw = o.optString("status", "")
            val st = if (stRaw.isEmpty()) null else stRaw.uppercase()
            QrPayResultJson(
                serverTxnId = o.optString("serverTxnId").ifEmpty { null },
                vehicleRegNo = o.optString("vehicleRegNo").ifEmpty { null },
                amountINR = o.optNullableDouble("amountINR"),
                newBalanceINR = o.optNullableDouble("newBalanceINR"),
                authCode = o.optString("authCode").ifEmpty { null },
                txnTime = o.optString("txnTime").ifEmpty { null },
                status = if (st == "FAILED") "FAILED" else "SUCCESS",
                quantityKg = o.optNullableDouble("quantityKg"),
            )
        }

    suspend fun driverGetTransactions(
        token: String,
        vehicleId: String,
        page: Int = 0,
    ): Result<TxnsPageParsed> =
        runCatchingSuspend {
            val q = if (page > 0) "?page=$page" else ""
            val vid = URLEncoder.encode(vehicleId, Charsets.UTF_8.name())
            normalizeDriverTransactionsPayload(unwrapDriverBodyJson(getRaw("/api/v0/driver-app/vehicles/$vid/transactions$q", token)))
        }

    private suspend inline fun <T> runCatchingSuspend(crossinline block: suspend () -> T): Result<T> =
        withContext(Dispatchers.IO) {
            try {
                Result.success(block())
            } catch (e: DriverApiException) {
                Result.failure(e)
            } catch (e: Exception) {
                Result.failure(e)
            }
        }

    private suspend fun getRaw(path: String, bearer: String?): String =
        withContext(Dispatchers.IO) {
            val url = "$base$path"
            val rb =
                Request.Builder()
                    .url(url)
                    .get()
                    .header("Accept", "application/json")
                    .header("User-Agent", USER_AGENT)
            bearer?.takeIf { it.isNotBlank() }?.let { rb.header("Authorization", "Bearer $it") }
            client.newCall(rb.build()).execute().use { resp ->
                val text = resp.body?.string().orEmpty()
                if (!resp.isSuccessful) failHttp(text, resp.code)
                text
            }
        }

    private suspend fun postJsonRaw(
        path: String,
        json: JSONObject,
        bearer: String?,
    ): String =
        withContext(Dispatchers.IO) {
            val body = json.toString().toRequestBody(JSON_MEDIA)
            val rb =
                Request.Builder()
                    .url("$base$path")
                    .post(body)
                    .header("Accept", "application/json")
                    .header("Content-Type", "application/json")
                    .header("User-Agent", USER_AGENT)
            bearer?.takeIf { it.isNotBlank() }?.let { rb.header("Authorization", "Bearer $it") }
            client.newCall(rb.build()).execute().use { resp ->
                val text = resp.body?.string().orEmpty()
                if (!resp.isSuccessful) failHttp(text, resp.code)
                text
            }
        }

    private fun failHttp(body: String, code: Int): Nothing {
        val msg =
            try {
                extractFleetApiErrorMessage(parseJsonFlexible(body))
            } catch (_: Exception) {
                null
            }
        throw DriverApiException(msg ?: "HTTP $code")
    }

    private fun requireJsonBody(text: String): Any? = unwrapDriverBody(parseJsonFlexible(text))

    private fun foListFromAny(data: Any?): List<FoListEntry> =
        when (data) {
            is JSONArray ->
                List(data.length()) { i ->
                    data.getJSONObject(i).let { fo ->
                        FoListEntry(
                            foCompanyId = fo.optLong("foCompanyId"),
                            foName = fo.optString("foName"),
                            foStatus = fo.optString("foStatus"),
                        )
                    }
                }
            else -> emptyList()
        }

    private fun parseHome(o: JSONObject): DriverHomeJson =
        DriverHomeJson(
            hasActiveVehicle = o.optBoolean("hasActiveVehicle"),
            foName = o.optNullableString("foName"),
            vehicleRegNo = o.optNullableString("vehicleRegNo"),
            vehicleId = o.optNullableString("vehicleId"),
            assignmentType = o.optNullableString("assignmentType"),
            isCurrentlyEligible = if (o.has("isCurrentlyEligible")) o.optBoolean("isCurrentlyEligible") else null,
            currentlyEligible = if (o.has("currentlyEligible")) o.optBoolean("currentlyEligible") else null,
            totalBalanceINR = o.optNullableDouble("totalBalanceINR"),
            shiftDaysOfWeek = o.optNullableString("shiftDaysOfWeek"),
            shiftStartTime = o.optNullableString("shiftStartTime"),
            shiftEndTime = o.optNullableString("shiftEndTime"),
            tripDate = o.optNullableString("tripDate"),
            tripStartTime = o.optNullableString("tripStartTime"),
            tripEndTime = o.optNullableString("tripEndTime"),
            tripStartLocation = o.optNullableString("tripStartLocation"),
        )

    private fun parseProfile(o: JSONObject): DriverProfileJson =
        DriverProfileJson(
            driverId = o.optString("driverId", ""),
            name = o.optString("name", ""),
            maskedMobile = o.optNullableString("maskedMobile"),
        )

    private fun parseAssignments(data: Any?): List<DriverAssignmentJson> =
        when (data) {
            is JSONArray ->
                List(data.length()) { i ->
                    data.getJSONObject(i).let { a ->
                        DriverAssignmentJson(
                            vehicleDriverId = a.optLong("vehicleDriverId"),
                            vehicleId = a.optString("vehicleId"),
                            vehicleRegNo = a.optString("vehicleRegNo"),
                            assignmentType = a.optString("assignmentType"),
                            status = a.optString("status"),
                            requiresPairing = a.optBoolean("requiresPairing"),
                            shiftDaysOfWeek = a.optNullableString("shiftDaysOfWeek"),
                            shiftStartTime = a.optNullableString("shiftStartTime"),
                            shiftEndTime = a.optNullableString("shiftEndTime"),
                            tripDate = a.optNullableString("tripDate"),
                            tripStartTime = a.optNullableString("tripStartTime"),
                            tripEndTime = a.optNullableString("tripEndTime"),
                            tripStartLocation = a.optNullableString("tripStartLocation"),
                            assignedAt = a.optNullableString("assignedAt"),
                        )
                    }
                }
            else -> emptyList()
        }

    private fun parseJsonFlexibleObj(unwrapped: Any?): Any? =
        when (unwrapped) {
            is String ->
                try {
                    parseJsonFlexible(unwrapped)
                } catch (_: Exception) {
                    unwrapped
                }
            else -> unwrapped
        }

    private fun requireJsonObject(any: Any?): JSONObject {
        require(any is JSONObject) { "Expected JSON object" }
        return any
    }

    private fun JSONObject.optNullableString(key: String): String? {
        if (!has(key) || JSONObject.NULL == opt(key)) return null
        return optString(key).ifBlank { null }
    }

    private fun JSONObject.optNullableDouble(key: String): Double? {
        if (!has(key) || JSONObject.NULL == opt(key)) return null
        return optDouble(key).let { if (it.isNaN()) null else it }
    }

    companion object {
        private val JSON_MEDIA = "application/json; charset=utf-8".toMediaType()
        private const val USER_AGENT = "mgl-fleet-android-sdk/1.0"
    }
}
